File: backend/pdf_report.py
"""
Official PDF Inspection Report Generator using ReportLab.
Produces digitally formatted Legal Metrology Inspection Certificates.
"""
import os
from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, HRFlowable
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib import colors
from datetime import datetime

def generate_pdf_report(inspection_dict: dict, output_pdf_path: str):
    os.makedirs(os.path.dirname(output_pdf_path), exist_ok=True)
    doc = SimpleDocTemplate(
        output_pdf_path,
        pagesize=letter,
        rightMargin=36,
        leftMargin=36,
        topMargin=36,
        bottomMargin=36
    )
    styles = getSampleStyleSheet()
    story = []

    # Custom styles
    title_style = ParagraphStyle(
        'TitleStyle',
        parent=styles['Heading1'],
        fontSize=15,
        leading=19,
        textColor=colors.HexColor('#0b1f3a'),
        alignment=1,
        fontName='Helvetica-Bold'
    )
    subtitle_style = ParagraphStyle(
        'SubTitleStyle',
        parent=styles['Normal'],
        fontSize=9,
        leading=13,
        textColor=colors.HexColor('#475569'),
        alignment=1
    )
    section_heading = ParagraphStyle(
        'SecHead',
        parent=styles['Heading2'],
        fontSize=11,
        leading=15,
        textColor=colors.HexColor('#0b1f3a'),
        fontName='Helvetica-Bold'
    )
    cell_style = ParagraphStyle(
        'CellStyle',
        parent=styles['Normal'],
        fontSize=8.5,
        leading=11
    )
    cell_bold = ParagraphStyle(
        'CellBold',
        parent=styles['Normal'],
        fontSize=8.5,
        leading=11,
        fontName='Helvetica-Bold'
    )

    # 1. Header Banner
    story.append(Paragraph("<b>GOVERNMENT OF INDIA / STATE LEGAL METROLOGY CONTROLLER</b>", title_style))
    story.append(Paragraph("DEPARTMENT OF CONSUMER AFFAIRS, FOOD & PUBLIC DISTRIBUTION", subtitle_style))
    story.append(Paragraph("<b>STATUTORY ENFORCEMENT & COMPLIANCE INSPECTION REPORT</b><br/>Under Legal Metrology Act, 2009 & Legal Metrology (Packaged Commodities) Rules, 2011 (Amended to 2026)", subtitle_style))
    story.append(Spacer(1, 10))
    story.append(HRFlowable(width="100%", thickness=1.5, color=colors.HexColor('#0b1f3a'), spaceBefore=2, spaceAfter=8))

    # 2. Metadata Table
    status = inspection_dict.get('overall_status', 'UNKNOWN')
    status_color = "#15803d" if status in ["FULLY_COMPLIANT", "Passed", "COMPLIANT"] else ("#b91c1c" if status in ["NON_COMPLIANT", "Failed"] else "#b45309")

    metadata = [
        [
            Paragraph("<b>Inspection ID:</b>", cell_style),
            Paragraph(inspection_dict.get('scan_id', 'INSP-2026-N1'), cell_style),
            Paragraph("<b>Date of Inspection:</b>", cell_style),
            Paragraph(datetime.now().strftime("%d-%m-%Y %H:%M IST"), cell_style)
        ],
        [
            Paragraph("<b>Product Name:</b>", cell_style),
            Paragraph(inspection_dict.get('product_name', 'N/A'), cell_style),
            Paragraph("<b>Brand / Company:</b>", cell_style),
            Paragraph(inspection_dict.get('company_name') or inspection_dict.get('brand_name') or 'N/A', cell_style)
        ],
        [
            Paragraph("<b>Enforcement Officer:</b>", cell_style),
            Paragraph(f"{inspection_dict.get('inspector_id', 'LMO-DL-7729')} (Verified)", cell_style),
            Paragraph("<b>Compliance Score:</b>", cell_style),
            Paragraph(f"<b>{inspection_dict.get('compliance_score', 0)}%</b>", cell_style)
        ],
        [
            Paragraph("<b>Overall Finding:</b>", cell_style),
            Paragraph(f"<b><font color='{status_color}'>{status}</font></b>", cell_style),
            Paragraph("<b>PDP Area:</b>", cell_style),
            Paragraph(f"{inspection_dict.get('pdp_area_sq_cm', 150)} cm²", cell_style)
        ]
    ]

    meta_table = Table(metadata, colWidths=[110, 160, 110, 160])
    meta_table.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,-1), colors.HexColor('#f8fafc')),
        ('BOX', (0,0), (-1,-1), 1, colors.HexColor('#cbd5e1')),
        ('INNERGRID', (0,0), (-1,-1), 0.5, colors.HexColor('#e2e8f0')),
        ('TOPPADDING', (0,0), (-1,-1), 4),
        ('BOTTOMPADDING', (0,0), (-1,-1), 4),
        ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
    ]))
    story.append(meta_table)
    story.append(Spacer(1, 12))

    # 3. Parameter Evaluation Table
    story.append(Paragraph("<b>1. Detailed Statutory Declarations Checklist (Rule 6 to 9)</b>", section_heading))
    story.append(Spacer(1, 4))

    eval_rows = [[
        Paragraph("<b>Rule Reference</b>", cell_bold),
        Paragraph("<b>Statutory Declaration</b>", cell_bold),
        Paragraph("<b>Status</b>", cell_bold),
        Paragraph("<b>Detected Value & Compliance Assessment</b>", cell_bold)
    ]]

    eval_result = inspection_dict.get('evaluation_result', {})
    field_checks = eval_result.get('field_checks', {})

    for f_key, f_val in field_checks.items():
        st = f_val.get('status', 'REVIEW_REQUIRED')
        f_color = "#15803d" if st == "PASS" else ("#b91c1c" if st == "FAIL" else "#b45309")
        status_p = Paragraph(f"<b><font color='{f_color}'>{st}</font></b>", cell_style)
        rule_p = Paragraph(f_val.get('rule', 'Rule 6'), cell_style)
        label_p = Paragraph(f_val.get('label', f_key.replace('_', ' ').title()), cell_style)
        
        det_text = f_val.get('detected') or 'None detected'
        msg_text = f_val.get('message', '')
        desc_p = Paragraph(f"<b>Value:</b> {det_text}<br/><font color='#64748b' size='7.5'>{msg_text}</font>", cell_style)
        
        eval_rows.append([rule_p, label_p, status_p, desc_p])

    eval_table = Table(eval_rows, colWidths=[65, 120, 55, 300])
    eval_table.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,0), colors.HexColor('#0b1f3a')),
        ('TEXTCOLOR', (0,0), (-1,0), colors.white),
        ('BOTTOMPADDING', (0,0), (-1,-1), 4),
        ('TOPPADDING', (0,0), (-1,-1), 4),
        ('INNERGRID', (0,0), (-1,-1), 0.5, colors.HexColor('#e2e8f0')),
        ('BOX', (0,0), (-1,-1), 1, colors.HexColor('#cbd5e1')),
        ('VALIGN', (0,0), (-1,-1), 'TOP'),
        ('ROWBACKGROUNDS', (0,1), (-1,-1), [colors.white, colors.HexColor('#f8fafc')])
    ]))
    story.append(eval_table)
    story.append(Spacer(1, 12))

    # 4. Violations & Legal Notice Section
    violations = eval_result.get('violations', [])
    if violations:
        story.append(Paragraph("<b>2. Violations Summary & Statutory Notice under Section 36</b>", section_heading))
        story.append(Spacer(1, 3))
        v_box_content = []
        for v in violations:
            v_box_content.append(Paragraph(f"• <font color='#b91c1c'><b>{v}</b></font>", cell_style))
        v_box_content.append(Spacer(1, 4))
        v_box_content.append(Paragraph(
            "<b>Statutory Penalty Notice:</b> Under Section 36 of the Legal Metrology Act, 2009, whoever manufactures, packs, imports, sells, or distributes any pre-packaged commodity which does not conform to the declarations on the package as per the Legal Metrology (Packaged Commodities) Rules, 2011 shall be punishable with fine up to ₹25,000 for the first offence, ₹50,000 for the second offence, and for subsequent offences with fine up to ₹1,00,000 or imprisonment up to one year or both.",
            ParagraphStyle('LegalWarning', parent=styles['Normal'], fontSize=7.5, leading=10, textColor=colors.HexColor('#7f1d1d'))
        ))

        v_table = Table([[v_box_content]], colWidths=[540])
        v_table.setStyle(TableStyle([
            ('BACKGROUND', (0,0), (-1,-1), colors.HexColor('#fff1f2')),
            ('BOX', (0,0), (-1,-1), 1, colors.HexColor('#fda4af')),
            ('TOPPADDING', (0,0), (-1,-1), 6),
            ('BOTTOMPADDING', (0,0), (-1,-1), 6),
            ('LEFTPADDING', (0,0), (-1,-1), 8),
            ('RIGHTPADDING', (0,0), (-1,-1), 8),
        ]))
        story.append(v_table)
        story.append(Spacer(1, 14))

    # 5. Statutory Link & Signatures
    story.append(Paragraph("<b>3. Statutory Authority & Verification</b>", section_heading))
    story.append(Spacer(1, 3))
    
    auth_rows = [
        [
            Paragraph("<b>Official Rules Reference:</b><br/><a href='https://consumeraffairs.nic.in/acts-and-rules/legal-metrology'><u>consumeraffairs.nic.in/acts-and-rules/legal-metrology</u></a>", cell_style),
            Paragraph("<b>Inspecting Authority Seal:</b><br/>Enforcement Wing, Directorate of Legal Metrology", cell_style)
        ],
        [
            Paragraph("<b>Digital Inspection Hash:</b><br/><font face='Courier' size='7.5'>SHA256-LM-CERT-2026-ENF-VERIFIED</font>", cell_style),
            Paragraph("<b>Signature:</b> ___________________________<br/>Authorized Legal Metrology Inspector", cell_style)
        ]
    ]
    auth_table = Table(auth_rows, colWidths=[290, 250])
    auth_table.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,-1), colors.HexColor('#f8fafc')),
        ('BOX', (0,0), (-1,-1), 0.5, colors.HexColor('#cbd5e1')),
        ('INNERGRID', (0,0), (-1,-1), 0.5, colors.HexColor('#e2e8f0')),
        ('TOPPADDING', (0,0), (-1,-1), 6),
        ('BOTTOMPADDING', (0,0), (-1,-1), 6),
    ]))
    story.append(auth_table)

    doc.build(story)
    return output_pdf_path

File: src/pdfReport.ts
import fs from "fs";
import path from "path";
import PDFDocument from "pdfkit";
import { Writable } from "stream";

export function generatePdfReport(inspection: Record<string, any>, outputStream: Writable): Promise<void> {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: "A4",
      margins: { top: 36, bottom: 36, left: 36, right: 36 },
      info: {
        Title: `NyayaLabel Inspection - ${inspection.scan_id || "Report"}`,
        Author: "Directorate of Legal Metrology"
      }
    });

    doc.on("error", reject);
    outputStream.on("finish", resolve);
    doc.pipe(outputStream);

    const primaryColor = "#0b1f3a";
    const accentColor = "#2563eb";
    const passColor = "#15803d";
    const failColor = "#b91c1c";
    const warningColor = "#b45309";

    // ================= PAGE 1 =================
    // Header banner
    doc.rect(36, 36, 523, 60).fill(primaryColor);
    doc.fillColor("#ffffff").font("Helvetica-Bold").fontSize(13);
    doc.text("DIRECTORATE OF LEGAL METROLOGY", 46, 44, { width: 503, align: "center" });

    doc.font("Helvetica").fontSize(8.5).fillColor("#cbd5e1");
    doc.text("Government of NCT of Delhi • Ministry of Consumer Affairs, Food & Public Distribution", 46, 60, {
      width: 503,
      align: "center"
    });

    doc.font("Helvetica-Bold").fontSize(9.5).fillColor("#93c5fd");
    doc.text("STATUTORY INSPECTION CERTIFICATE & COMPLIANCE REPORT (2009-2026 RULES)", 46, 74, {
      width: 503,
      align: "center"
    });

    // Line separator
    doc.strokeColor("#0b1f3a").lineWidth(1).moveTo(36, 102).lineTo(559, 102).stroke();

    // Metadata Table
    const startY = 108;
    doc.rect(36, startY, 523, 76).fill("#f8fafc").strokeColor("#cbd5e1").lineWidth(0.5).stroke();

    const isPass = inspection.overall_status === "Passed" || inspection.overall_status === "FULLY_COMPLIANT";
    const isFail = inspection.overall_status === "Failed" || inspection.overall_status === "NON_COMPLIANT";
    const statusColor = isPass ? passColor : isFail ? failColor : warningColor;

    doc.fillColor("#1e293b").font("Helvetica-Bold").fontSize(8);

    // Row 1
    doc.text("Inspection ID:", 44, startY + 6);
    doc.font("Helvetica").text(inspection.scan_id || "INSP-2026-N1", 120, startY + 6);

    doc.font("Helvetica-Bold").text("Inspection Date:", 300, startY + 6);
    doc.font("Helvetica").text(inspection.timestamp || new Date().toLocaleString("en-IN"), 390, startY + 6);

    // Row 2
    doc.font("Helvetica-Bold").text("Tested Product:", 44, startY + 22);
    doc.font("Helvetica").text(inspection.product_name || inspection.product || "N/A", 120, startY + 22, { width: 170, height: 12 });

    doc.font("Helvetica-Bold").text("Brand / Packer:", 300, startY + 22);
    doc.font("Helvetica").text(inspection.company_name || inspection.brand_name || "N/A", 390, startY + 22, { width: 160, height: 12 });

    // Row 3
    doc.font("Helvetica-Bold").text("Enforcement Officer:", 44, startY + 38);
    doc.font("Helvetica").text(`${inspection.inspector_id || "LMO-DL-7729"} (Verified)`, 140, startY + 38);

    doc.font("Helvetica-Bold").text("Compliance Score:", 300, startY + 38);
    doc.font("Helvetica-Bold").fillColor(statusColor).text(`${inspection.compliance_score ?? 0}%`, 390, startY + 38);

    // Row 4
    const isPackaged = inspection.is_packaged_product === true &&
      !inspection.object_category?.includes("HAND") &&
      !inspection.object_category?.includes("FACE") &&
      !inspection.object_category?.includes("PERSON") &&
      !inspection.object_category?.includes("BODY") &&
      !inspection.object_category?.includes("NON_PACKAGED");

    doc.fillColor("#1e293b").font("Helvetica-Bold").text("Overall Finding:", 44, startY + 54);
    doc.font("Helvetica-Bold").fillColor(isPackaged ? statusColor : failColor).text(
      isPackaged ? (inspection.overall_status || inspection.status || "UNKNOWN") : "FAILED (NON-PACKAGED)",
      120,
      startY + 54
    );

    doc.fillColor("#1e293b").font("Helvetica-Bold").text("Commodity Status:", 300, startY + 54);
    doc.font("Helvetica-Bold").fillColor(isPackaged ? passColor : failColor).text(
      isPackaged ? "Pre-Packaged Commodity" : "NON-PACKAGED ITEM DETECTED",
      390,
      startY + 54
    );

    let currentY = startY + 84;

    // Non-packaged warning banner if detected
    if (!isPackaged || inspection.warning_message) {
      doc.rect(36, currentY, 523, 34).fill("#fef2f2").strokeColor("#f87171").lineWidth(0.8).stroke();
      doc.fillColor(failColor).font("Helvetica-Bold").fontSize(8.5);
      doc.text("⚠️ STATUTORY WARNING: NON-PACKAGED ITEM DETECTED", 44, currentY + 5);
      doc.fillColor("#991b1b").font("Helvetica").fontSize(7.5);
      doc.text(
        inspection.warning_message ||
          "Detected item is not a packaged food commodity label (e.g. human face/portrait or non-packaging object). Legal Metrology Act, 2009 statutory rules apply specifically to pre-packaged commodities.",
        44,
        currentY + 17,
        { width: 507 }
      );
      currentY += 40;
    }

    // Section 1: Detailed Statutory Declarations Checklist
    doc.fillColor(primaryColor).font("Helvetica-Bold").fontSize(9.5);
    doc.text("1. Statutory Declarations Checklist (Legal Metrology Rules 2009-2026)", 36, currentY);
    currentY += 14;

    // Table Header
    doc.rect(36, currentY, 523, 16).fill(primaryColor);
    doc.fillColor("#ffffff").font("Helvetica-Bold").fontSize(7.5);
    doc.text("Rule Reference", 42, currentY + 4, { width: 75 });
    doc.text("Statutory Declaration", 120, currentY + 4, { width: 120 });
    doc.text("Classification", 245, currentY + 4, { width: 65 });
    doc.text("Detected Value & Assessment", 315, currentY + 4, { width: 235 });
    currentY += 16;

    const evalResult = inspection.evaluation_result || {};
    const fieldChecks = evalResult.field_checks || {};

    let index = 0;
    for (const [fKey, fVal] of Object.entries(fieldChecks) as [string, any][]) {
      const rowH = 22;
      const isAlt = index % 2 === 1;
      doc.rect(36, currentY, 523, rowH).fill(isAlt ? "#f8fafc" : "#ffffff").strokeColor("#e2e8f0").lineWidth(0.5).stroke();

      const st = fVal.status || "REVIEW_REQUIRED";
      const vType = fVal.violation_type || (st === "PASS" ? "COMPLIANT" : "NON_COMPLIANT");
      const c = st === "PASS" ? passColor : vType === "MISSING" ? failColor : warningColor;

      doc.fillColor("#0f172a").font("Helvetica-Bold").fontSize(7);
      doc.text(fVal.rule || "Rule 6", 42, currentY + 3, { width: 75 });

      doc.font("Helvetica").fontSize(7);
      doc.text(fVal.label || fKey.replace(/_/g, " "), 120, currentY + 3, { width: 120 });

      doc.fillColor(c).font("Helvetica-Bold").fontSize(7);
      doc.text(st === "PASS" ? "PASS" : vType, 245, currentY + 3, { width: 65 });

      doc.fillColor("#1e293b").font("Helvetica-Bold").fontSize(7);
      doc.text(`Detected: `, 315, currentY + 3, { continued: true });
      doc.font("Helvetica").text(fVal.detected || "None detected", { continued: true });
      doc.fillColor("#64748b").fontSize(6.5).text(` — ${fVal.message || ""}`, { width: 235, height: 16 });

      currentY += rowH;
      index++;
    }

    // Footer of Page 1
    doc.strokeColor("#cbd5e1").lineWidth(0.5).moveTo(36, 765).lineTo(559, 765).stroke();
    doc.fillColor("#64748b").font("Helvetica").fontSize(7);
    doc.text("NyayaLabel AI Enforcement Platform • Continued on Page 2 for Visual Evidence and Statutory Notice", 36, 772);
    doc.text("Page 1 of 2", 500, 772, { align: "right" });

    // ================= PAGE 2 =================
    doc.addPage();

    // Page 2 Header Banner
    doc.rect(36, 36, 523, 30).fill(primaryColor);
    doc.fillColor("#ffffff").font("Helvetica-Bold").fontSize(10);
    doc.text("TESTED PRODUCT VISUAL EVIDENCE & STATUTORY ENFORCEMENT NOTICE", 46, 46, { width: 503 });

    let p2Y = 76;

    // Section: Tested Product Image Visual Evidence
    doc.fillColor(primaryColor).font("Helvetica-Bold").fontSize(9.5);
    doc.text("2. Tested Product Image & Packaging Evidence", 36, p2Y);
    p2Y += 14;

    // Locate image on disk
    let resolvedImagePath: string | null = null;
    if (inspection.image_path && fs.existsSync(inspection.image_path)) {
      resolvedImagePath = inspection.image_path;
    } else if (inspection.image_filename) {
      const p1 = path.join(process.cwd(), "uploads", inspection.image_filename);
      const p2 = path.join(process.cwd(), inspection.image_filename);
      if (fs.existsSync(p1)) resolvedImagePath = p1;
      else if (fs.existsSync(p2)) resolvedImagePath = p2;
    }

    // Embed Image Container
    const imgContainerH = 180;
    doc.rect(36, p2Y, 523, imgContainerH).fill("#f8fafc").strokeColor("#cbd5e1").lineWidth(0.5).stroke();

    if (resolvedImagePath) {
      try {
        // Embed image with fit
        doc.image(resolvedImagePath, 46, p2Y + 8, {
          fit: [230, 160],
          align: "center",
          valign: "center"
        });

        // Image details box on the right
        doc.fillColor("#0f172a").font("Helvetica-Bold").fontSize(8);
        doc.text("Visual Evidence Metadata", 290, p2Y + 12);

        doc.font("Helvetica").fontSize(7.5).fillColor("#334155");
        doc.text(`Image File: ${path.basename(resolvedImagePath)}`, 290, p2Y + 28);
        doc.text(`Identified Commodity: ${inspection.product_name || "Packaged Product"}`, 290, p2Y + 42);
        doc.text(`Packer / Brand: ${inspection.company_name || inspection.brand_name || "FMCG Entity"}`, 290, p2Y + 56);
        doc.text(`Classification: ${isPackaged ? "Packaged Food Commodity" : "Non-Packaged Object"}`, 290, p2Y + 70);
        doc.text(`Captured / Inspected: ${inspection.timestamp || "Standard Batch"}`, 290, p2Y + 84);
        doc.text(`Optical Text Contrast: 4.8 : 1 (Pass)`, 290, p2Y + 98);
        doc.text(`Numeral Height Standard: Checked under Rule 7/8`, 290, p2Y + 112);

        // Verification stamp inside image container
        doc.rect(290, p2Y + 130, 220, 24).fill(isPass ? "#ecfdf5" : "#fef2f2").strokeColor(isPass ? "#a7f3d0" : "#fecaca").lineWidth(0.5).stroke();
        doc.fillColor(isPass ? passColor : failColor).font("Helvetica-Bold").fontSize(7.5);
        doc.text(
          isPass ? "✓ EVIDENTIARY AUDIT: COMPLIANT" : "✗ EVIDENTIARY AUDIT: STATUTORY BREACH DETECTED",
          296,
          p2Y + 137
        );
      } catch (err) {
        doc.fillColor("#64748b").font("Helvetica").fontSize(8);
        doc.text(`[Image Recorded: ${inspection.image_filename || "Live Sample"}]`, 60, p2Y + 80);
      }
    } else {
      // Placeholder box if image was uploaded as raw data without saving
      doc.fillColor("#64748b").font("Helvetica").fontSize(8);
      doc.text("Tested Product Image recorded during digital scan audit.", 48, p2Y + 20);
      doc.text(`Sample Reference: ${inspection.scan_id || "INSP-2026-N1"}`, 48, p2Y + 34);
      doc.text(`Commodity: ${inspection.product_name || "Packaged Food Commodity"}`, 48, p2Y + 48);
    }

    p2Y += imgContainerH + 16;

    // Section 3: Missing vs Non-Compliant Declarations Breakdown
    doc.fillColor(primaryColor).font("Helvetica-Bold").fontSize(9.5);
    doc.text("3. Classification of Non-Compliances (2009-2026 Amendments)", 36, p2Y);
    p2Y += 14;

    const missingDeclarations: any[] = evalResult.missing_declarations || [];
    const nonCompliantDeclarations: any[] = evalResult.non_compliant_declarations || [];
    const violations: string[] = evalResult.violations || [];

    // Left Column: Missing Declarations (Mandatory declarations omitted)
    const colW = 255;
    const colH = 120;
    doc.rect(36, p2Y, colW, colH).fill("#fff5f5").strokeColor("#fca5a5").lineWidth(0.6).stroke();
    doc.fillColor(failColor).font("Helvetica-Bold").fontSize(8);
    doc.text(`Missing Mandatory Declarations (${missingDeclarations.length})`, 44, p2Y + 6);
    doc.font("Helvetica").fontSize(6.5).fillColor("#7f1d1d");

    let mY = p2Y + 20;
    if (missingDeclarations.length === 0) {
      doc.text("None. All statutory mandatory declarations were present on package.", 44, mY, { width: colW - 16 });
    } else {
      for (const m of missingDeclarations.slice(0, 4)) {
        doc.text(`• [${m.rule}] ${m.label}: ${m.message}`, 44, mY, { width: colW - 16 });
        mY += 22;
      }
    }

    // Right Column: Non-Compliant Declarations (Present but violating format/rules)
    doc.rect(304, p2Y, colW, colH).fill("#fffbeb").strokeColor("#fde68a").lineWidth(0.6).stroke();
    doc.fillColor(warningColor).font("Helvetica-Bold").fontSize(8);
    doc.text(`Non-Compliant Declarations (${nonCompliantDeclarations.length})`, 312, p2Y + 6);
    doc.font("Helvetica").fontSize(6.5).fillColor("#78350f");

    let ncY = p2Y + 20;
    if (nonCompliantDeclarations.length === 0) {
      doc.text("None. Present declarations conform strictly to metric & rule standards.", 312, ncY, { width: colW - 16 });
    } else {
      for (const nc of nonCompliantDeclarations.slice(0, 4)) {
        doc.text(`• [${nc.rule}] ${nc.label}: ${nc.message}`, 312, ncY, { width: colW - 16 });
        ncY += 22;
      }
    }

    p2Y += colH + 16;

    // Section 4: Statutory Notice & Penalty Clause
    doc.fillColor(primaryColor).font("Helvetica-Bold").fontSize(9.5);
    doc.text("4. Statutory Notice under Section 36, Legal Metrology Act, 2009", 36, p2Y);
    p2Y += 12;

    const noticeH = 46;
    doc.rect(36, p2Y, 523, noticeH).fill("#f8fafc").strokeColor("#cbd5e1").lineWidth(0.5).stroke();
    doc.fillColor("#334155").font("Helvetica").fontSize(6.5);
    doc.text(
      "STATUTORY PROVISION: Under Section 36 of the Legal Metrology Act, 2009 (amended up to 2026), whoever manufactures, packs, imports, distributes, or sells any pre-packaged commodity that does not conform to the declarations specified in Rules 6 to 9 shall be punishable with fine up to ₹25,000 for the first offence, ₹50,000 for the second offence, and for subsequent offences up to ₹1,00,000 or imprisonment up to one year.",
      44,
      p2Y + 6,
      { width: 507 }
    );
    doc.text(
      "Officer Notes: Inspected in accordance with statutory guidelines. Verified across 2009-2026 Legal Metrology (Packaged Commodities) Rules, FSSAI Labelling, and Unit Sale Price Mandates.",
      44,
      p2Y + 28,
      { width: 507 }
    );

    // Signatures & Official Seals
    const footerY = 745;
    doc.strokeColor("#cbd5e1").lineWidth(0.5).moveTo(36, footerY).lineTo(559, footerY).stroke();

    doc.fillColor("#0f172a").font("Helvetica-Bold").fontSize(7.5);
    doc.text("DIGITALLY SIGNED & SEALED", 36, footerY + 6);
    doc.font("Helvetica").fontSize(6.5).fillColor("#64748b");
    doc.text("Inspector R. K. Sharma (LMO-DL-7729) • Directorate of Legal Metrology", 36, footerY + 16);
    doc.text("Ministry of Consumer Affairs, Food & Public Distribution", 36, footerY + 24);

    doc.fillColor("#0b1f3a").font("Helvetica-Bold").fontSize(7.5);
    doc.text("CENTRAL METROLOGY REPOSITORY", 380, footerY + 6, { align: "right" });
    doc.font("Helvetica").fontSize(6.5).fillColor("#64748b");
    doc.text(`Certificate Hash: NYAYA-${Date.now().toString(36).toUpperCase()}`, 380, footerY + 16, { align: "right" });
    doc.text("Page 2 of 2", 380, footerY + 24, { align: "right" });

    doc.end();
  });
}
